xline/src/storage/compact/mod.rs

```diff
@@ -1,19 +1,100 @@
 use std::{sync::Arc, time::Duration};
 
+use async_trait::async_trait;
+use curp::{client::Client, cmd::ProposeId, error::ProposeError};
 use event_listener::Event;
+use periodic_compactor::PeriodicCompactor;
+use revision_compactor::RevisionCompactor;
 use tokio::{sync::mpsc::Receiver, time::sleep};
+use utils::config::AutoCompactConfig;
+use uuid::Uuid;
 
 use super::{
     index::{Index, IndexOperate},
     storage_api::StorageApi,
     KvStore,
 };
+use crate::{
+    revision_number::RevisionNumberGenerator,
+    rpc::{CompactionRequest, RequestWithToken},
+    server::command::Command,
+};
+
+/// mod revision compactor;
+mod revision_compactor;
+
+/// mod periodic compactor;
+mod periodic_compactor;
 
 /// compact task channel size
 pub(crate) const COMPACT_CHANNEL_SIZE: usize = 32;
 
+/// Compactor trait definition
+#[async_trait]
+pub(crate) trait Compactor: std::fmt::Debug + Send + Sync {
+    /// run an auto-compactor
+    async fn run(&self);
+    /// pause an auto-compactor when the current node denotes to a non-leader role
+    fn pause(&self);
+    /// resume an auto-compactor when the current becomes a leader
+    fn resume(&self);
+}
+
+/// `Compactable` trait indicates a method that receives a given revision and proposes a compact proposal
+#[cfg_attr(test, mockall::automock)]
+#[async_trait]
+pub(crate) trait Compactable: std::fmt::Debug + Send + Sync {
+    /// do compact
+    async fn compact(&self, revision: i64) -> Result<(), ProposeError>;
+}
+
+#[async_trait]
+impl Compactable for Client<Command> {
+    async fn compact(&self, revision: i64) -> Result<(), ProposeError> {
+        let request = CompactionRequest {
+            revision,
+            physical: false,
+        };
+        let request_wrapper = RequestWithToken::new_with_token(request.into(), None);
+        let propose_id = ProposeId::new(format!("auto-compactor-{}", Uuid::new_v4()));
+        let cmd = Command::new(vec![], request_wrapper, propose_id);
+        let _cmd_res = self.propose(cmd).await?;
+        Ok(())
+    }
+}
+
+/// Boot up an auto-compactor background task.
+pub(crate) async fn auto_compactor(
+    is_leader: bool,
+    client: Arc<Client<Command>>,
+    revision_getter: Arc<RevisionNumberGenerator>,
+    shutdown_trigger: Arc<Event>,
+    auto_compact_cfg: AutoCompactConfig,
+) -> Arc<dyn Compactor> {
+    let auto_compactor: Arc<dyn Compactor> = match auto_compact_cfg {
+        AutoCompactConfig::Periodic(period) => {
+            PeriodicCompactor::new_arc(is_leader, client, revision_getter, shutdown_trigger, period)
+        }
+        AutoCompactConfig::Revision(retention) => RevisionCompactor::new_arc(
+            is_leader,
+            client,
+            revision_getter,
+            shutdown_trigger,
+            retention,
+        ),
```

Comment:
Can we get these two modes to work together

Reply:
I don't get the point of the "work together" here. Do you mean to enable both of them at the same time?

Reply:
Yes, I mean to enable both of them at the same time

Reply:
No, they are different strategies. You can only enable either periodic mode or revision mode at the same time.